sluice/udp_listener.go

```diff
@@ -0,0 +1,321 @@
+// replies is correct.
+func (ln *UDPListener) setSocketOptions() error {
+	// As we might be listening on an interface that supports both IPv4 and IPv6,
+	// try to set the socket options on both.
+	// Only report an error if it fails on both.
+	err4 := ipv4.NewPacketConn(ln.sock).SetControlMessage(ipv4.FlagDst|ipv4.FlagInterface, true)
+	err6 := ipv6.NewPacketConn(ln.sock).SetControlMessage(ipv6.FlagDst|ipv6.FlagInterface, true)
+	if err4 != nil && err6 != nil {
+		return err4
+	}
+
+	return nil
+}
+
+// UDPConn simulates a connection for a stateless protocol.
+type UDPConn struct {
+	ln      *UDPListener
+	addr    *net.UDPAddr
+	oob     []byte
+	closed  *abool.AtomicBool
+	closing chan struct{}
+
+	buf []byte
+	in  chan []byte
+
+	inactivityCnt *uint32
+}
+
+// Read reads data from the connection.
+// Read can be made to time out and return an error after a fixed
+// time limit; see SetDeadline and SetReadDeadline.
+func (conn *UDPConn) Read(b []byte) (n int, err error) {
+	// Check if connection is closed.
+	if conn.closed.IsSet() {
+		return 0, io.EOF
+	}
+
+	// Mark as active.
+	atomic.StoreUint32(conn.inactivityCnt, 0)
+
+	// Get new buffer.
+	if conn.buf == nil {
+		select {
+		case conn.buf = <-conn.in:
+			if conn.buf == nil {
+				return 0, io.EOF
+			}
+		case <-conn.closing:
+			return 0, io.EOF
+		}
+	}
+
+	// Serve from buffer.
+	copy(b, conn.buf)
+	if len(b) >= len(conn.buf) {
+		copied := len(conn.buf)
+		conn.buf = nil
+		return copied, nil
+	}
+	copied := len(b)
+	conn.buf = conn.buf[copied:]
+	return copied, nil
+}
+
+// Write writes data to the connection.
+// Write can be made to time out and return an error after a fixed
+// time limit; see SetDeadline and SetWriteDeadline.
+func (conn *UDPConn) Write(b []byte) (n int, err error) {
+	// Check if connection is closed.
+	if conn.closed.IsSet() {
+		return 0, io.EOF
+	}
+
+	// Mark as active.
+	atomic.StoreUint32(conn.inactivityCnt, 0)
+
+	n, _, err = conn.ln.sock.WriteMsgUDP(b, conn.oob, conn.addr)
+	return n, err
+}
+
+// Close is a no-op as UDP connections share a single socket. Just stop sending
+// packets without closing.
+func (conn *UDPConn) Close() error {
+	if conn.closed.SetToIf(false, true) {
+		close(conn.closing)
+	}
+	return nil
+}
+
+// LocalAddr returns the local network address.
+func (conn *UDPConn) LocalAddr() net.Addr {
+	return conn.ln.sock.LocalAddr()
+}
+
+// RemoteAddr returns the remote network address.
+func (conn *UDPConn) RemoteAddr() net.Addr {
+	return conn.addr
+}
+
+// SetDeadline is a no-op as UDP connections share a single socket.
+func (conn *UDPConn) SetDeadline(t time.Time) error {
+	return nil
+}
+
+// SetReadDeadline is a no-op as UDP connections share a single socket.
+func (conn *UDPConn) SetReadDeadline(t time.Time) error {
+	return nil
+}
+
+// SetWriteDeadline is a no-op as UDP connections share a single socket.
+func (conn *UDPConn) SetWriteDeadline(t time.Time) error {
+	return nil
+}
```

Comment:
Those should actually call through to `conn.ln.sock.XYZ` so the timeout behavior will work as one expects. 
Even if that would be a no-op because `net.UDPConn` does nothing it will ensure we pick up any changes in the stdlib as well.